internal/jobs/iocp.go

```diff
@@ -0,0 +1,58 @@
+package jobs
+
+import (
+	"sync"
+	"unsafe"
+
+	"github.com/Microsoft/hcsshim/internal/winapi"
+	"golang.org/x/sys/windows"
+)
+
+var (
+	ioInitOnce sync.Once
+	// Global iocp handle that will be re-used for every job object
+	ioCompletionPort windows.Handle
+	// Mapping of job handle to channel that that job object owns to be able to receive
+	// notifications.
+	jobMap = make(map[uint32]chan IOCPNotif)
+)
+
+// IOCPNotif represents a job object specific IOCP message.
+// Code will hold the value of the job object specific message that was received.
+// Key will hold the job object handle value so the job the message is for can be determined.
+// Err holds any error returned from polling the IOCP.
+type IOCPNotif struct {
+	Code uint32
+	Key  uint32
+	Err  error
+}
+
+// ioCompletionProcessor processes completed async IOs forever
+func pollIOCP(iocpHandle windows.Handle) {
+	var (
+		overlapped uintptr
+		code       uint32
+		key        uint32
+	)
+
+	for {
+		err := windows.GetQueuedCompletionStatus(iocpHandle, &code, &key, (**windows.Overlapped)(unsafe.Pointer(&overlapped)), windows.INFINITE)
```

Comment:
`GetQueuedCompletionStatus` taking a `*uint32` value for `key` seems like a bug. The `key` value attached for the job is a `uintptr`, so we could be corrupting memory when we receive `key` here.

I think we should switch to using a copy of `GetQueuedCompletionStatus` defined in `winapi`, and send a PR to fix this issue in `golang.org/x/sys/windows`.